### DISCUSSION TITLE
MultiTenancy UserRole Filter

### DISCUSSION THREAD

**Discussion by havvanur160401046** [serenity-is/Serenity#5505] [Help]
![image](https://user-images.githubusercontent.com/24252615/108478250-4eac0a00-72a5-11eb-9344-595c82ce546c.png)
I applied multiple tenants in my project, but roles are not filtered while assigning user roles, all roles in the database are coming, how can I do this?
![image](https://user-images.githubusercontent.com/24252615/108478283-58357200-72a5-11eb-8742-c5820f75ce29.png)
I guess I'll do a filter here, how can I do it

**Comment by WasimAhmad** [comment]
have you implemented MultiTenancy for RoleRow? show me the code for Rolerow.cs

**Reply by WasimAhmad** [reply]
here roles are shown using standard row lookupscript. you have to write a custom one.

```csharp
    [LookupScript("Administration.RoleLookup", Expiration = -1)]
    public class RoleLookupScript : RowLookupScript<RoleRow>
    {
        private RoleRow.RowFields fld = RoleRow.Fields;
        public RoleLookupScript()
        {
            IdField = fld.RoleId.PropertyName;
            TextField = fld.RoleName.PropertyName;
        }

        protected override void PrepareQuery(SqlQuery query)
        {
            query
                .Select(fld.RoleId)
                .SelectAs($"{fld.RoleName.Expression} || ' (' || {fld.Module.Expression} || ')'", fld.RoleName)
                //.Select(fld.RoleName)
                .Select(fld.Module);

            if (!Authorization.Username.Equals("admin"))
            {
// Apply your multitendency filter here...
                var permissions = new UserPermissionRepository().getAssignedModules();
                query.Where(fld.Module.In(permissions));
            }
        }
    }
}

// and finally update RoleRow.cs like this

    [LookupScript(typeof(RoleLookupScript))]
    public sealed class RoleRow : Row, IIdRow, INameRow, IUnitRow
    {.....

**Reply by havvanur160401046** [reply]
![image](https://user-images.githubusercontent.com/24252615/108673505-d471c480-74f4-11eb-9a7c-138c8765d12b.png)
roles are working right only "kisi" has to come and it feels right
but when assigning roles to the user, all roles come
![image](https://user-images.githubusercontent.com/24252615/108673696-29153f80-74f5-11eb-8a21-8e9dd80f33b0.png)
In which file should I create the lookupscript you have given?

**Reply by havvanur160401046** [reply]
RoleRow.cs
` [LookupScript("Administration.Role")]
    public sealed class RoleRow : Row, IIdRow, INameRow, IMultiTenantRow
    {
        [DisplayName("Role Id"), Identity, ForeignKey("Roles", "RoleId"), LeftJoin("jRole")]
        public Int32? RoleId
        {
            get { return Fields.RoleId[this]; }
            set { Fields.RoleId[this] = value; }
        }

        [DisplayName("Role Name"), Size(100), NotNull, QuickSearch]
        public String RoleName
        {
            get { return Fields.RoleName[this]; }
            set { Fields.RoleName[this] = value; }
        }

        [Insertable(false), Updatable(false)]
        public Int32? TenantId
        {
            get { return Fields.TenantId[this]; }
            set { Fields.TenantId[this] = value; }
        }

        IIdField IIdRow.IdField
        {
            get { return Fields.RoleId; }
        }

        StringField INameRow.NameField
        {
            get { return Fields.RoleName; }
        }

        public Int32Field TenantIdField
        {
            get { return Fields.TenantId; }
        }`

**Reply by WasimAhmad** [reply]
RoleLookupScript.cs

**Reply by havvanur160401046** [reply]
I solved the problem with a few changes in the code you wrote
Thank you:)

**Reply by TechAspirant** [reply]
@WasimAhmad Can you please share the Multitenancy example in .Net5 as I am facing issues in migrating multitenancy example to .Net5.

**Reply by axelcasauran** [reply]
I have the same issue and I tried the above workaround but no luck. Can anyone help us?
